first version of cloudcompserdiff CLI tool to compare two composer env's

ant-laz opened the pull request:
## Description

First version of cloudcomposerdiff CLI tool to highlight the differences between two composer environments

## Checklist
- [x] I have followed [Sample Guidelines from AUTHORING_GUIDE.MD](https://github.com/GoogleCloudPlatform/python-docs-samples/blob/main/AUTHORING_GUIDE.md)
- [x] README is updated to include [all relevant information](https://github.com/GoogleCloudPlatform/python-docs-samples/blob/main/AUTHORING_GUIDE.md#readme-file)
- [ ] **Tests** pass:   `nox -s py-3.6` (see [Test Environment Setup](https://github.com/GoogleCloudPlatform/python-docs-samples/blob/main/AUTHORING_GUIDE.md#test-environment-setup))
- [x] **Lint** pass:   `nox -s lint` (see [Test Environment Setup](https://github.com/GoogleCloudPlatform/python-docs-samples/blob/main/AUTHORING_GUIDE.md#test-environment-setup))
- [ ] These samples need a new **API enabled** in testing projects to pass (let us know which ones)
- [ ] These samples need a new/updated **env vars** in testing projects set to pass (let us know which ones)
- [x] Please **merge** this PR for me once it is approved.
- [ ] This sample adds a new sample directory, and I updated the [CODEOWNERS file](https://github.com/GoogleCloudPlatform/python-docs-samples/blob/main/.github/CODEOWNERS) with the codeowners for this sample

ant-laz committed:
first version of cloudcompserdiff CLI tool to compare two composer env's

ant-laz committed:
adds missing licence headers

leahecole commented on cloudcomposerdiff/.gitignore:
```diff
@@ -0,0 +1,4 @@
+__pycache__/
+*.pyc
+venv/
+*.egg-info
```
Please remove this file - it's great to have locally but not needed in the repo 🙂

ant-laz replied:
I looked into this, but was unable to do so. 

Could you please tell me how to do this? 

Doesn't git need to know about .gitignore to ignore these files ?

leahecole replied:
Instead of removing the file, I decided to just remove line 3, which is what's specific to your environment. I think it's fair that the other lines would want to be ignored by someone when running/maintaining this tool and are not unique to your specific env

leahecole commented on cloudcomposerdiff/requirements.txt:
```diff
@@ -0,0 +1,63 @@
+#
+# This file is autogenerated by pip-compile with python 3.10
```
Ooo, interesting. I'm not familiar with using pip-compile and am not against it, but have a few questions and comments

1. Question - If I'm understanding this correctly, this has all dependencies implicit and explicit?
2. Question/comment - in this repo, I had planned to use [Mend renovate](https://www.mend.io/free-developer-tools/renovate/) for automated dependency updates. In other repos, I have it look at `requirements.txt`. It seems like that would be inappropriate here because it may get the implicit dependencies out of sync with their parents. Is the single source of truth to watch here the `pyproject.toml`?
3. comment - If the source of truth is the `pyproject.toml`, we should probably consider pinning dependencies to a version or a range of versions. I am not 100% sure the best practices with this and am open to your opinions here, or, I can solicit opinions from other folks in the Python community
4. questions - How often do you run pip-compile? From a maintenance perspective, it seems like I would want to add it as part of a build/test pipeline whenever a dependency is changed, right?

ant-laz replied:
I discovered pip-compile through this article: https://realpython.com/pypi-publish-python-package/

I will try to answer your questions here:

1. yes, implicit & explicit dependencies
2. May be easier to re-jig things on my end to fit y our Med renovate tooling
3. Please get some other opinions from python folks with more packaging experience. happy to pin versions based on their advice
4. yes, it'd be great if this were automated. I often forget to do this at the moment because it's a manual process of me executing the command in the CONTRIBUTING.md guide

Leaving this comment open to capture advice of you & your colleagues on how to proceed.

leahecole commented on cloudcomposerdiff/setup.py:
```diff
@@ -0,0 +1,23 @@
+ limitations under the License.
+ """
+
+# this file is required to make editable installs work
+# ongoing work to remove need for this file:
+# https://github.com/pypa/setuptools/issues/2816
```
Do you think that editable installs will be a dealbreaker for users? Would you like to publish this to PyPI instead? (I would have to double check whether or not I have to do a turbolaunch for it but I do know how to do PyPI packages)

ant-laz replied:
As per README & CONTRIBUTING I am only suggested editable installs for future developers. 

For regular end-users I am just recommending local PyPI installs, e.g. clone repo & get pip to install from source code.

We discussed avoiding PyPi due to concerns about who will support this tool in the long term.

I'll leave this comment open to hear what you & your colleagues advise as the best way forward here.

leahecole commented on cloudcomposerdiff/pyproject.toml:
```diff
@@ -0,0 +1,31 @@
+name = "cloudcomposerdiff"
+version = "1.0.0"
+description = "Determine the diff between two cloud composer environments"
+dependencies = [
```
I made a note of this in requirements.txt, but depending on what best practice is, we might want to pin to a dependency version or a range of versions.

ant-laz replied:
Understood. 

I'll leave this comment open to hear the best way forward generally in terms of packaging.

leahecole commented on cloudcomposerdiff/pyproject.toml:
```diff
@@ -0,0 +1,31 @@
+[project]
+name = "cloudcomposerdiff"
+version = "1.0.0"
+description = "Determine the diff between two cloud composer environments"
```
```suggestion
description = "Determine the diff between two Cloud Composer environments"
```

leahecole commented on cloudcomposerdiff/src/cloudcomposerdiff/lib/__init__.py:
```diff
@@ -0,0 +1,15 @@
+"""
+ Copyright 2022 Google LLC
+
+ Licensed under the Apache License, Version 2.0 (the "License");
+ you may not use this file except in compliance with the License.
+ You may obtain a copy of the License at
+
+      https://www.apache.org/licenses/LICENSE-2.0
+
+ Unless required by applicable law or agreed to in writing, software
+ distributed under the License is distributed on an "AS IS" BASIS,
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License.
+ """
```
```suggestion
```

__init__.py is exempt from license headers

ant-laz replied:
Thanks, I didn't know this !

leahecole commented on cloudcomposerdiff/src/cloudcomposerdiff/__init__.py:
```diff
@@ -0,0 +1,18 @@
+"""
+ Copyright 2022 Google LLC
+
+ Licensed under the Apache License, Version 2.0 (the "License");
+ you may not use this file except in compliance with the License.
+ You may obtain a copy of the License at
+
+      https://www.apache.org/licenses/LICENSE-2.0
+
+ Unless required by applicable law or agreed to in writing, software
+ distributed under the License is distributed on an "AS IS" BASIS,
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License.
+ """
+
+# Version o of the cloudcomposerdiff package
+__version__ = "1.0.0"
```
```suggestion
```

This is a suggestion I'm open to discussing but here's my assumptions and why I'm proposing just removing this altogether:

- We have the version of the package in the `pyproject.toml` which is the primary spot where I'd expect it to be updated.
-  When it's in this file, is it auto-updated as part of a release process? Or is it manual? If it's manual, it's annoying to put it in two places
- are there major advantages to having the version as part of the initialization code? (There might be and if so, please tell me so I can learn!!) I'm assuming no which is why I'm proposing its removal but I could be totally wrong

ant-laz replied:
No auto-update. 

As per my mail, I followed this guide to packing up my code: https://realpython.com/pypi-publish-python-package/

As per these instructions: https://github.com/ant-laz/composer-utilities/blob/main/cloudcomposerdiff/CONTRIBUTING.md#update-the-version-number-of-the-package . bumpver is used to update version numbers anywhere.

Please tell me how I should be setting this up. I am out-of-my-depth here ! 

If I only have to specify version in pyproject.toml, great ! 

I'll leave this comment open to hear your view on the best way forward re. packaging & versioning.

leahecole commented on cloudcomposerdiff/src/cloudcomposerdiff/__main__.py:
```diff
@@ -0,0 +1,111 @@
+@click.option(
+    "--env1_project_id",
+    type=str,
+    help="GCP Project ID for Cloud Composer Environment #1",
```
```suggestion
    help="Project ID for Cloud Composer Environment #1",
```

leahecole commented on cloudcomposerdiff/src/cloudcomposerdiff/__main__.py:
```diff
@@ -0,0 +1,111 @@
+@click.option(
+    "--env1_location",
+    type=str,
+    help="GCP location for Cloud Composer Environment #1",
```
```suggestion
    help="Compute Engine region for Cloud Composer Environment #1",
```

^I changed it to region, but if it's zone, specify that it's zone instead

leahecole commented on cloudcomposerdiff/src/cloudcomposerdiff/__main__.py:
```diff
@@ -0,0 +1,111 @@
+@click.option(
+    "--env1_name",
+    type=str,
+    help="GCP name for Cloud Composer Environment #1",
```
```suggestion
    help="Name for Cloud Composer Environment #1",
```

leahecole commented on cloudcomposerdiff/src/cloudcomposerdiff/__main__.py:
```diff
@@ -0,0 +1,111 @@
+@click.option(
+    "--env2_project_id",
+    type=str,
+    help="GCP Project ID for Cloud Composer Environment #2",
```
```suggestion
    help="Project ID for Cloud Composer Environment #2",
```

leahecole commented on cloudcomposerdiff/src/cloudcomposerdiff/__main__.py:
```diff
@@ -0,0 +1,111 @@
+@click.option(
+    "--env2_location",
+    type=str,
+    help="GCP location for Cloud Composer Environment #2",
```
```suggestion
    help="Compute Engine region for Cloud Composer Environment #2",
```

^ if it's zone, change to zone instead

leahecole commented on cloudcomposerdiff/src/cloudcomposerdiff/__main__.py:
```diff
@@ -0,0 +1,111 @@
+@click.option(
+    "--env2_name",
+    type=str,
+    help="GCP name for Cloud Composer Environment #2",
```
```suggestion
    help="Name for Cloud Composer Environment #2",
```

leahecole commented on cloudcomposerdiff/src/cloudcomposerdiff/lib/comparator.py:
```diff
@@ -0,0 +1,60 @@
+        console.print(table)
+
+    def output_diffs_as_json(self) -> None:
+        pass
```
Why is this a `pass`?

ant-laz replied:
I simply had not had the time ! 

I have implemented this now.

I have also extended the README.md to mention this functionalty

leahecole commented on cloudcomposerdiff/src/cloudcomposerdiff/lib/difference.py:
```diff
@@ -0,0 +1,30 @@
+ See the License for the specific language governing permissions and
+ limitations under the License.
+ """
+
+
```
I don't necessarily know if this merits a change, but I may seek another opinion here. Does this need to be its own class in its own file if all it is is initialization? (The answer might be yes! But it also might be no 😁 )

ant-laz replied:
Fair point. 

Happy to take your direction here. 

Will leave this comment open.

leahecole commented on cloudcomposerdiff/src/cloudcomposerdiff/lib/service.py:
```diff
@@ -0,0 +1,33 @@
+from google.cloud.orchestration.airflow import service_v1
+
+
+class GCPComposerService:
+    def __init__(self, project_id: str, location: str, name: str) -> None:
+        self.client: service_v1.EnvironmentsClient = service_v1.EnvironmentsClient()
+        self.name = f"projects/{project_id}/locations/{location}/environments/{name}"
+
+    def fetch_env(self) -> service_v1.types.Environment:
+        request: service_v1.GetEnvironmentRequest = service_v1.GetEnvironmentRequest(
+            name=self.name
+        )
+        response: service_v1.types.Environment = self.client.get_environment(
+            request=request
+        )
+        return response
```
Hmm, does this need to be an entire class? It seems like it's mostly repeated logic from the client library right? Or am I missing an extension of library logic here?

If it really is just repeated logic from the client library, creating the client and making a request seems like it's part of the diff operation imho. 

Happy again to be challenged with this logic though

ant-laz replied:
You're right in that there is no extension here of service_v1.EnvironmentsClient.

I do this for three reasons:

1. Reduce code duplication. At a minimum I need to repeat these steps for two composer environments. Putting them in a class was a neat way of avoid code repetition.
2. Reduce memory usage. Per environment, I create just a single object in memory. I pass a reference to this object around to each "diff strategy" which reads from the objects attributes. This avoids making a environment object per diff.
3. Reduce network calls. Per environment, I make a single network call to fetch its attributes. This avoid making a separate network call per diff strategy.

I'll leave this comment open to hear your advice on how to proceed here.

leahecole commented on cloudcomposerdiff/tests/test_diffenvimage.py:
```diff
@@ -0,0 +1,49 @@
+from cloudcomposerdiff.lib.strategies.diffenvimage import DiffEnvImage
+
+
+def test_diff_env_image_strategy():
```
Just double checking - this doesn't actually make a service call right? it just constructs the object and uses your comparison?

ant-laz replied:
yes, that's exactly right ! 

I avoid making a service call by hand creating two "service_v1.types.Environment" instances. 

I then pass these instances directly into the DiffEnvImage instance to test its detect_difference method.

leahecole commented on cloudcomposerdiff/src/cloudcomposerdiff/lib/strategies/diffairflowconfig.py:
```diff
@@ -0,0 +1,100 @@
+        diffs = []
+        if config1 == {} and config2 == {}:
+            # both environments do not have any airflow config overrides
+            pass
```
nit - instead of passing is it worth logging something useful to the end user? (Or will that logging happen somewhere else?)
context: I absolutely try to avoid using `pass` unless absolutely necessary - I firmly believe that in 99% of cases a pass can be avoided with a useful info level log 😁

ant-laz replied:
Ok, fair point! 

I have avoided using a pass and have added some basic logging here to raise attention to the fact the the supplied composer environments have no airflow config overrides set.

leahecole commented on cloudcomposerdiff/src/cloudcomposerdiff/lib/strategies/diffairflowconfig.py:
```diff
@@ -0,0 +1,100 @@
+"""
```
nit please rename file with snake_case (`diff_airflow_config.py`)

ant-laz replied:
done.

leahecole commented on cloudcomposerdiff/src/cloudcomposerdiff/lib/strategies/diffenvimage.py:
```diff
@@ -0,0 +1,38 @@
+"""
```
nit - please rename file to `diff_env_image.py`

ant-laz replied:
done.

leahecole commented on cloudcomposerdiff/tests/test_diffenvimage.py:
```diff
@@ -0,0 +1,49 @@
+"""
```
nit - please rename to `test_diff_env_image.py`

ant-laz replied:
done.

leahecole commented on cloudcomposerdiff/src/cloudcomposerdiff/lib/strategies/diffenvimage.py:
```diff
@@ -0,0 +1,38 @@
+ See the License for the specific language governing permissions and
+ limitations under the License.
+ """
+
```
So I understand the logic that the other file has - it's looking for differences in the Airflow config. Some thoughts about this diff

1. Is this logic separate because it's considered a Composer attribute versus Airflow attributes?
2. Why is image_version the only thing checked? What about other configs?

ant-laz replied:
As per my mail, I am using the "strategy" design pattern.

I have an abstract base class called [EnvironmentAttributeDiffer](https://github.com/ant-laz/composer-utilities/blob/main/cloudcomposerdiff/src/cloudcomposerdiff/lib/strategies/strategy.py) that describe a generic strategy to finding differences between two composer environments.

[DiffEnvImage](https://github.com/ant-laz/composer-utilities/blob/main/cloudcomposerdiff/src/cloudcomposerdiff/lib/strategies/diffenvimage.py) and  [DiffAirflowConfig](https://github.com/ant-laz/composer-utilities/blob/main/cloudcomposerdiff/src/cloudcomposerdiff/lib/strategies/diffairflowconfig.py) inherit from this abstract base class and provide a concrete implementations of the  detect_difference method.

The [compare_environments](https://github.com/ant-laz/composer-utilities/blob/main/cloudcomposerdiff/src/cloudcomposerdiff/lib/comparator.py#L35) method of the EnvironmentComparator class takes as input any EnvironmentAttributeDiffer as a "strategy" to diff with.

This way the app can have lots of different "strategies" to diff two environments, they can all just be plugged into the existing code as per [main](https://github.com/ant-laz/composer-utilities/blob/main/cloudcomposerdiff/src/cloudcomposerdiff/__main__.py#L89-L90).

This approach makes it easier to add new strategies in the future. Just add a new class to the strategy [folder](https://github.com/ant-laz/composer-utilities/tree/main/cloudcomposerdiff/src/cloudcomposerdiff/lib/strategies) & add a reference in [main](https://github.com/ant-laz/composer-utilities/blob/main/cloudcomposerdiff/src/cloudcomposerdiff/__main__.py#L89-L90). Easy ! 

I am open to better approaches to designing the tool to make it easy to extend !

leahecole commented on cloudcomposerdiff/README.md:
```diff
@@ -0,0 +1,184 @@
+# A python command line tool to diff two Cloud composer environments.
+
```
```suggestion
# cloudcomposerdiff
A python command line tool to compare two Cloud Composer environments.

```

ant-laz replied:
Thanks for the suggestion. I have made this change.

leahecole commented on cloudcomposerdiff/README.md:
```diff
@@ -0,0 +1,184 @@
+# A python command line tool to diff two Cloud composer environments.
+
+## Running this code
```
Are there prerequisites a user must have before running this? What assumptions are you making? You don't necessarily have to tell them how to do everything but I would say things like
- What version of Python is required
- What version of Composer you can use this on
- Any IAM permissions they would need on their project/Composer environment
- Any restrictions (can you run this across all regions and zones that Composer is available in? Can you run this with all networking configuirations?)

ant-laz replied:
Good point I have added a Prerequisites section to cover these points, thanks !

leahecole commented on cloudcomposerdiff/README.md:
```diff
@@ -0,0 +1,184 @@
+
+### Clone repository
+
+Clone the code from GitHub to your local machine
```
```suggestion
[Clone the code](https://docs.github.com/en/repositories/creating-and-managing-repositories/cloning-a-repository) from GitHub to your local machine
```

leahecole commented on cloudcomposerdiff/README.md:
````diff
@@ -0,0 +1,184 @@
+
+### Install locally with pip
+
+```
````
```suggestion
```shell
```

leahecole commented on cloudcomposerdiff/README.md:
````diff
@@ -0,0 +1,184 @@
+### Install locally with pip
+
+```
+pip install cloudcomposerdiff
````
```suggestion
cd composer-utilities
pip install cloudcomposerdiff
```

I'm assuming the user wants to be in the root directory of the utils

leahecole replied:
Actually, I tried this from the directory and I had to have it as `pip install cloudcomposerdiff/` is that what you are expecting?

ant-laz replied:
Yes, that's expected.

I assume that users are in the cloudcomposerdiff directory when running the pip install command. 

I have extended the README to make this clearer.

leahecole commented on cloudcomposerdiff/README.md:
```diff
@@ -0,0 +1,184 @@
+
+Select a user which has access to both composer environments to compare.
+
+Wit this user account, initialize gcloud:
```
```suggestion
With this user account, initialize gcloud:
```

leahecole commented on cloudcomposerdiff/README.md:
```diff
@@ -0,0 +1,184 @@
+
+Install [gcloud](https://cloud.google.com/sdk/gcloud)
+
+Select a user which has access to both composer environments to compare.
```
```suggestion
Select a user which has access to both Cloud Composer environments to compare.
```

Also, suggest adding the IAM roles in a prereq section then mentioning them again here

ant-laz replied:
good catch. I have added IAM roles here again as suggested.

leahecole commented on cloudcomposerdiff/README.md:
````diff
@@ -0,0 +1,184 @@
+
+Wit this user account, initialize gcloud:
+
+```
````
```suggestion
```shell
```

leahecole commented on cloudcomposerdiff/README.md:
````diff
@@ -0,0 +1,184 @@
+
+Next, authorize the application to access cloud composer environments:
+
+```
````
```suggestion
```shell
```

leahecole commented on cloudcomposerdiff/README.md:
````diff
@@ -0,0 +1,184 @@
+
+### Use the appplication
+
+```
````
```suggestion
```shell
```

leahecole commented on cloudcomposerdiff/README.md:
````diff
@@ -0,0 +1,184 @@
+gcloud auth application-default login
+```
+
+### Use the appplication
````
This section needs more info. What should the user expect to happen when they run this command? (how long will it take? minutes? hours? days? What happens when it runs?) Tell them here

ant-laz replied:
Thank you for the feedback. I have added several videos (in the form of GIFs) to the README.   These should clearly explain to users what will happen when they run the tool.  Please let me know if the GIFs are not clear.

leahecole commented on cloudcomposerdiff/README.md:
````diff
@@ -0,0 +1,184 @@
+### Use the appplication
+
+```
+cloudcomposerdiff \
+--env1_project_id <> \
+--env1_location <> \
+--env1_name <> \
+--env2_project_id <> \
+--env2_location <> \
+--env2_name <>
````
```suggestion
cloudcomposerdiff \
--env1_project_id YOUR_PROJECT_ID \
--env1_location YOUR_ENV1_LOCATION \
--env1_name YOUR_ENV1_NAME \
--env2_project_id YOUR_PROJECT_ID \
--env2_location YOUR_ENV2_LOCATION \
--env2_name YOUR_ENV2_NAME
```

leahecole commented on cloudcomposerdiff/README.md:
````diff
@@ -0,0 +1,184 @@
+--env2_location <> \
+--env2_name <>
+```
+
````
include an example command for users and what expected output would look like. If expected output is too verbose to put here, lmk and we can brainstorm what to do. 

Example of an example command
```shell
cloudcomposerdiff \
--env1_project_id my-project \
--env1_location us-central1 \
--env1_name my-environment-1 \
--env2_project_id my-project \
--env2_location us-east1 \
--env2_name my-environment-2
```

ant-laz replied:
As per my other comment, I have uploaded 3 videos (in the form of GIFs) which demonstrate using the tool. The videos show what the input looks likes and what the example output looks like (in tabular form). Please let me know if these videos are not clear & I need to add more details to the README

leahecole commented on cloudcomposerdiff/README.md:
````diff
@@ -0,0 +1,184 @@
+```
+
+## Developing this application
+
````
Add prereqs here like version of python

ant-laz replied:
Good point. 

I have added a prereqs section here as well.

leahecole commented on cloudcomposerdiff/README.md:
````diff
@@ -0,0 +1,184 @@
+Install [pyenv](https://github.com/pyenv/pyenv).
+
+Use pyenv to install a suitable version of python, e.g. 3.10.0
+```
````
```suggestion
```shell
```

leahecole commented on cloudcomposerdiff/README.md:
````diff
@@ -0,0 +1,184 @@
+```
+
+Use pyenv to switch to python 3.10.0
+```
````
```suggestion
```shell
```

leahecole commented on cloudcomposerdiff/README.md:
````diff
@@ -0,0 +1,184 @@
+
+Use pyenv to install a suitable version of python, e.g. 3.10.0
+```
+$ pyenv install 3.10.0    
````
```suggestion
pyenv install 3.10.0    
```

leahecole commented on cloudcomposerdiff/README.md:
````diff
@@ -0,0 +1,184 @@
+
+Use pyenv to switch to python 3.10.0
+```
+$ pyenv shell 3.10.0
````
```suggestion
pyenv shell 3.10.0
```
not important additional nit besides removing $ - Would it be worth using `local` instead of `shell`? Either works, just curious.

ant-laz replied:
Thanks for catching the "$", I have removed it

pyenv local creates a .python-version file in the current directory. 

To improve developer experience, let's avoid the creation of hidden files like this and stick to pyenv shell.

leahecole commented on cloudcomposerdiff/README.md:
````diff
@@ -0,0 +1,184 @@
+
+### Create a virtual environment with venv module
+
+```
````
```suggestion
```shell
```

leahecole commented on cloudcomposerdiff/README.md:
````diff
@@ -0,0 +1,184 @@
+### Create a virtual environment with venv module
+
+```
+python3 -m venv venv
````
```suggestion
python -m venv venv
```

If you're using pyenv, I believe it's okay to use `python` here instead of `python3`. I try to avoid using `python3` because it implies that `python2` is around and I don't want to give anyone any ideas about using it 😁

ant-laz replied:
ah ok, thanks for the feedback. 

Yes, agree. Let's not bring up python2 ! 

I have changed all references from python3 -> python.

leahecole commented on cloudcomposerdiff/README.md:
````diff
@@ -0,0 +1,184 @@
+```
+
+activate the virtual environment
+```
````
```suggestion
```shell
```

leahecole commented on cloudcomposerdiff/README.md:
````diff
@@ -0,0 +1,184 @@
+```
+source venv/bin/activate
+```
+check that we are using the python of the virtual environment
````
```suggestion
check that we are using the Python of the virtual environment
```

leahecole commented on cloudcomposerdiff/README.md:
````diff
@@ -0,0 +1,184 @@
+```
+(venv ) $ python3 -m pip install --editable .
+```
+Install additiona dev only dependencies that end-users do not get
````
```suggestion
Install additional dev only dependencies that end-users do not get
```

leahecole commented on cloudcomposerdiff/README.md:
````diff
@@ -0,0 +1,184 @@
+--env2_name <>
+```
+
+## Developing this application
````
I think that from here down this should be moved to a CONTRIBUTING.md or MAINTAINING.md file so that end users don't have to worry about it

ant-laz replied:
great suggestion!  I have done this.

leahecole commented on cloudcomposerdiff/README.md:
````diff
@@ -0,0 +1,184 @@
+(venv ) $ python3 -m pip install --editable .
+```
+Install additiona dev only dependencies that end-users do not get
+```
````
```suggestion
```shell
```

leahecole commented on cloudcomposerdiff/README.md:
````diff
@@ -0,0 +1,184 @@
+
+with the above done, running the package will include code changes. This can
+be used for testing during development.
+```
+cloudcomposerdiff \
+--env1_project_id <> \
+--env1_location <> \
+--env1_name <> \
+--env2_project_id <> \
+--env2_location <> \
+--env2_name <>
+```
````
This command is included here without much else in the name of instruction. Should it be here? If so, add a quick note about why you'd run it and the expected output, and please refactor it to reflect the suggestion I made earlier that removes the brackets

ant-laz replied:
Good point ! 

I have added some more text here to explain what's going on.

As per my other comments, I have also added 3 GIFs to hopefully make the concept of the tool super clear !

leahecole commented on cloudcomposerdiff/README.md:
````diff
@@ -0,0 +1,184 @@
+### Run tests w/ pytest
+
+Navigate to the directory containing pyproject.toml
+```
````
```suggestion
```shell
```

leahecole commented on cloudcomposerdiff/README.md:
````diff
@@ -0,0 +1,184 @@
+
+Apply the sort to all files recurisvely
+
+```
````
```suggestion
```shell
```

leahecole commented on cloudcomposerdiff/README.md:
````diff
@@ -0,0 +1,184 @@
+```
+
+### Sort imports with isosort
+
````
for my learning purposes, why do we do this?

ant-laz replied:
PEP-8 has a convention for imports in python: https://peps.python.org/pep-0008/#imports . The python module isort https://pypi.org/project/isort/ does the heavy lifting of converting imports in the code I write into the PEP-8 convention of grouping imports by type. Let me know if this does not make sense.

leahecole commented on cloudcomposerdiff/README.md:
````diff
@@ -0,0 +1,184 @@
+
+Apply the formatter to all files recurisvely
+
+```
````
```suggestion
```shell
```

leahecole commented on cloudcomposerdiff/README.md:
````diff
@@ -0,0 +1,184 @@
+
+Apply the formatter to all files recurisvely
+
+```
````
```suggestion
```shell
```

leahecole commented on cloudcomposerdiff/README.md:
````diff
@@ -0,0 +1,184 @@
+Do not manually update requirements.txt
+
+Use a pip-tools command to sync pyproject.toml to requirements.txt
+```
````
```suggestion
```shell
```

leahecole commented on cloudcomposerdiff/README.md:
````diff
@@ -0,0 +1,184 @@
+
+E.g. to preview an increase to minor 
+
+```
````
```suggestion
```shell
```

leahecole commented on cloudcomposerdiff/README.md:
````diff
@@ -0,0 +1,184 @@
+
+E.g. to actually increase minor 
+
+```
````
```suggestion
```shell
```

leahecole reviewed (changes requested):
First of all YAY thank you for being my first util onboarded and for bearing with me for this experiment!! A few things to note

Notes for you

1.  Annoying nit - We've dropped "GCP" from external branding. I tried to make suggestions where I saw it but I do think you'll need to rename a class
2. Is there a design pattern you're following for this? I'm asking because I haven't done object-oriented Python in awhile and I'm trying to gain more insight into the organization of the code - in particular, things that are in the `lib` directory. My initial impression is the organization feels very Java-y :) 
3. I see a test for diffenvimage, but not one for diffairflowconfig. Did you maybe miss a file, or do you plan to add it?
4. Are all of your test dependencies (anything the tests need but just regular running the service doesn't need) included in pyproject.toml or somewhere? I want to know where I need to look for dependencies for running tests
5. Thanks for using type hints!!!! :) 

FYIs
1. I need to add a lint check and I'll try to implement that in the next couple of days
2. I need to add something that can run your tests as part of a presubmit check. I will also work on adding that in the next couple of days
3. The README was the last thing I reviewed so if I ask questions that seem like they're answered in the README that might be why (thinking about the requirements question in particular)

ant-laz committed:
adds example as a GIF to readme to better explain value of tool

ant-laz committed:
fixes typo in relative paths for GIFs

ant-laz committed:
added GIF examples of lots of matches & zero matches, added them to README

ant-laz committed:
Update cloudcomposerdiff/README.md

Co-authored-by: Leah E. Cole <6719667+leahecole@users.noreply.github.com>

ant-laz committed:
Update cloudcomposerdiff/README.md

Co-authored-by: Leah E. Cole <6719667+leahecole@users.noreply.github.com>

ant-laz committed:
Update cloudcomposerdiff/README.md

Co-authored-by: Leah E. Cole <6719667+leahecole@users.noreply.github.com>

ant-laz committed:
Update cloudcomposerdiff/README.md

Co-authored-by: Leah E. Cole <6719667+leahecole@users.noreply.github.com>

ant-laz committed:
Update cloudcomposerdiff/README.md

Co-authored-by: Leah E. Cole <6719667+leahecole@users.noreply.github.com>

ant-laz committed:
Update cloudcomposerdiff/README.md

Co-authored-by: Leah E. Cole <6719667+leahecole@users.noreply.github.com>

ant-laz committed:
Update cloudcomposerdiff/README.md

Co-authored-by: Leah E. Cole <6719667+leahecole@users.noreply.github.com>

ant-laz committed:
Update cloudcomposerdiff/src/cloudcomposerdiff/lib/__init__.py

Co-authored-by: Leah E. Cole <6719667+leahecole@users.noreply.github.com>

ant-laz committed:
Update cloudcomposerdiff/pyproject.toml

Co-authored-by: Leah E. Cole <6719667+leahecole@users.noreply.github.com>

ant-laz committed:
Update cloudcomposerdiff/src/cloudcomposerdiff/__main__.py

Co-authored-by: Leah E. Cole <6719667+leahecole@users.noreply.github.com>

ant-laz committed:
Update cloudcomposerdiff/src/cloudcomposerdiff/__main__.py

Co-authored-by: Leah E. Cole <6719667+leahecole@users.noreply.github.com>

ant-laz committed:
Update cloudcomposerdiff/src/cloudcomposerdiff/__main__.py

Co-authored-by: Leah E. Cole <6719667+leahecole@users.noreply.github.com>

ant-laz committed:
Update cloudcomposerdiff/src/cloudcomposerdiff/__main__.py

Co-authored-by: Leah E. Cole <6719667+leahecole@users.noreply.github.com>

ant-laz committed:
Update cloudcomposerdiff/src/cloudcomposerdiff/__main__.py

Co-authored-by: Leah E. Cole <6719667+leahecole@users.noreply.github.com>

ant-laz committed:
Update cloudcomposerdiff/src/cloudcomposerdiff/__main__.py

Co-authored-by: Leah E. Cole <6719667+leahecole@users.noreply.github.com>

ant-laz committed:
Update cloudcomposerdiff/README.md

Co-authored-by: Leah E. Cole <6719667+leahecole@users.noreply.github.com>

ant-laz committed:
Update cloudcomposerdiff/README.md

Co-authored-by: Leah E. Cole <6719667+leahecole@users.noreply.github.com>

ant-laz committed:
Update cloudcomposerdiff/README.md

Co-authored-by: Leah E. Cole <6719667+leahecole@users.noreply.github.com>

ant-laz committed:
Update cloudcomposerdiff/README.md

Co-authored-by: Leah E. Cole <6719667+leahecole@users.noreply.github.com>

ant-laz committed:
Update cloudcomposerdiff/README.md

Co-authored-by: Leah E. Cole <6719667+leahecole@users.noreply.github.com>

ant-laz committed:
Update cloudcomposerdiff/README.md

Co-authored-by: Leah E. Cole <6719667+leahecole@users.noreply.github.com>

ant-laz committed:
Update cloudcomposerdiff/README.md

Co-authored-by: Leah E. Cole <6719667+leahecole@users.noreply.github.com>

ant-laz committed:
Update cloudcomposerdiff/README.md

Co-authored-by: Leah E. Cole <6719667+leahecole@users.noreply.github.com>

ant-laz committed:
Update cloudcomposerdiff/README.md

Co-authored-by: Leah E. Cole <6719667+leahecole@users.noreply.github.com>

ant-laz committed:
Update cloudcomposerdiff/README.md

Co-authored-by: Leah E. Cole <6719667+leahecole@users.noreply.github.com>

ant-laz committed:
Update cloudcomposerdiff/README.md

Co-authored-by: Leah E. Cole <6719667+leahecole@users.noreply.github.com>

ant-laz committed:
Update cloudcomposerdiff/README.md

Co-authored-by: Leah E. Cole <6719667+leahecole@users.noreply.github.com>

ant-laz committed:
Update cloudcomposerdiff/README.md

Co-authored-by: Leah E. Cole <6719667+leahecole@users.noreply.github.com>

ant-laz committed:
Update cloudcomposerdiff/README.md

Co-authored-by: Leah E. Cole <6719667+leahecole@users.noreply.github.com>

ant-laz committed:
Update cloudcomposerdiff/README.md

Co-authored-by: Leah E. Cole <6719667+leahecole@users.noreply.github.com>

ant-laz committed:
updated readme with feedback from review

ant-laz committed:
README.md is for end-users, CONTRIBUTING.md is for developers wishing to extend the tool

ant-laz committed:
removing developer instructions from README.md

ant-laz committed:
added a Prerequisites section

ant-laz committed:
extends gcloud instructions in README

ant-laz committed:
added text to better explain how to execute the program

ant-laz committed:
clearer installation instructions & contributing prerequisites

ant-laz committed:
renamed files based on feedback. added logging to avoid simplying doing a pass

ant-laz committed:
adds JSON export functionality

ant-laz committed:
fixed typo

ant-laz committed:
adds support to compare env variables & pypi packages

ant-laz committed:
updated README to list attributes that are compared

ant-laz committed:
fixed typo

leahecole commented on cloudcomposerdiff/.gitignore:
```diff
@@ -0,0 +1,4 @@
+__pycache__/
+*.pyc
+venv/
```
```suggestion
```

leahecole committed:
Update cloudcomposerdiff/.gitignore

Instead of removing, I decided to just remove what would be specific to your environment. In this case, the virtualenv

leahecole committed:
Merge branch 'main' into main

leahecole committed:
Merge branch 'main' into main

leahecole committed:
Merge branch 'main' into main

leahecole committed:
Merge branch 'main' into main

leahecole committed:
Merge branch 'main' into main

leahecole committed:
Merge branch 'main' into main

leahecole committed:
run black on tests/ directory

leahecole committed:
Merge branch 'main' into main

leahecole committed:
run isort with google profile

leahecole commented on cloudcomposerdiff/CONTRIBUTING.md:
````diff
@@ -0,0 +1,147 @@
+Apply the sort to all files recurisvely
+
+```shell
+isort .
````
```suggestion
isort . --profile=black
```

This will ensure that the import order complies with the flake8 lint check and with black (I made a real mess with figuring out the profiles earlier 😅 )

ant-laz replied:
Thanks Leah !

leahecole commented on cloudcomposerdiff/CONTRIBUTING.md:
````diff
@@ -0,0 +1,147 @@
+Apply the formatter to all files recurisvely
+
+```shell
+flake8 --max-line-length=88 src
````
```suggestion
flake8 . --show-source --builtin=gettext --max-complexity=29 --import-order-style=pep8 --ignore=E121,E123,E126,E203,E226,E24,E266,E501,E704,W503,W504,I202 --max-line-length=88
```

I know this is a huge amount of text but it's copied straight from the GH action https://github.com/GoogleCloudPlatform/composer-utilities/blob/4864e3ca553438e223b86e71e111221be3409574/.github/workflows/cloudcomposerdiff.yml#L46 (and derived from the samples lint that we use for cloud python samples)

leahecole replied:
note that users would also have to install two additional dependencies (`flake8-import-order` and `flake8-annotations`)

ant-laz replied:
Thank you for sorting all this out Leah ! 

I will update the dev dependencies of pyproject.toml to include flake8-import-order and flak8-annotations

leahecole committed:
experimet with black and isort

leahecole committed:
different import order style

leahecole committed:
Merge branch 'main' into main

leahecole committed:
Merge branch 'main' into main

leahecole committed:
Merge branch 'main' into main

leahecole committed:
Merge branch 'main' into main

leahecole committed:
Merge branch 'main' into main

leahecole committed:
Merge branch 'main' into main

leahecole commented on cloudcomposerdiff/README.md:
```diff
@@ -0,0 +1,115 @@
+
+As per the example above, this tool looks at various Cloud Composer environment
+attributes and diffs them. It shows which attributes have different values across
+environemnts, and what those values are. It also shows which attributes have a common
```
```suggestion
environments, and what those values are. It also shows which attributes have a common
```

leahecole commented on cloudcomposerdiff/README.md:
```diff
@@ -0,0 +1,115 @@
+
+The tool can compare the follow attributes of Cloud Composer environments:
+
+* Cloud Compser image version
+* AirFlow configuration overrides
+* PyPi packages
```
```suggestion
* Cloud Composer image version
* Airflow configuration overrides
* PyPI packages
```

ant-laz replied:
Thanks for cleaning this up !

leahecole commented on cloudcomposerdiff/README.md:
````diff
@@ -0,0 +1,115 @@
+from within the cloudcomposerdiff directory run the following command
+
+```shell
+pip install cloudcomposerdiff
````
Can you double check this command? I am not able to run it as is because right now this tells pip to look for a package in PyPI named `cloudcomposerdiff`

leahecole replied:
I think it needs to be `pip install .` or `pip install -r requirements.txt`

leahecole replied:
(I think it's `pip install .` because that looks for our setup.py)

leahecole replied:
cc @chalmerlowe for his wisdom here 😁

chalmerlowe replied:
Typically, if you are in the directory and want to do a local `pip install`, then it is customary to run `pip install .`.

Since the comment above indicates that this is intended to run the application (instead of installing in editable mode for development purposes) I would expect `pip install .` to be the appropriate answer.

leahecole replied:
thank you @chalmerlowe !!

ant-laz replied:
Thank @leahecole and @chalmerlowe for the feedback & direction here! 

I spun up a new VM, clone the repo and ran into the exact same issue you did. 

I've updated the README with this instruction.

leahecole committed:
Merge branch 'main' into main

leahecole committed:
Merge branch 'main' into main

ant-laz committed:
Update cloudcomposerdiff/CONTRIBUTING.md

Co-authored-by: Leah E. Cole <6719667+leahecole@users.noreply.github.com>

ant-laz committed:
Update cloudcomposerdiff/CONTRIBUTING.md

Co-authored-by: Leah E. Cole <6719667+leahecole@users.noreply.github.com>

ant-laz committed:
Update cloudcomposerdiff/README.md

Co-authored-by: Leah E. Cole <6719667+leahecole@users.noreply.github.com>

ant-laz committed:
Update cloudcomposerdiff/README.md

Co-authored-by: Leah E. Cole <6719667+leahecole@users.noreply.github.com>

ant-laz committed:
removes unneeded class wrapping the composer service

ant-laz committed:
zapping flake errors

ant-laz committed:
zapping flake8 errors using python 3.10 features

ant-laz committed:
applying black to files

ant-laz committed:
corrects installation instructions for end-users

leahecole committed:
Merge branch 'main' into main

leahecole reviewed (approved):
LGTM if it looks good to @chalmerlowe

leahecole commented:
snippet bot was previously passing and seems to be having a hard time org-wide, I am not considering it a blocker

chalmerlowe reviewed (approved):
I am gonna mark this LGTM with the following caveats:

* I made a cursory review of this (it is 24 files and well over a thousand lines of code)
* I am not familiar with Cloud Composer and how it works/what it does
* My lack of familiarity with Cloud Composer means I can not vouch for the adequacy of this implementation, etc.
